### PR TITLE
Flag for manual review design doc

### PR DESCRIPTION
**Pull request by bkjohnson**
This is a draft design doc for what we've been calling the automated "flag for manual review" process.  It is not _strictly_ a design doc since I ended up creating the solution ahead of time as I was experimenting with Github's API and Circle's CI config, but I tried to capture some design decisions and open questions.

All feedback welcome.

[Here is an earlier issue](https://github.com/department-of-veterans-affairs/va.gov-team/issues/2521) which describes some of the goals. I will try to work that into the design doc somehow.

### CODE REVIEW

File: platform/engineering/design-docs/flag-for-manual-review.md
Comment (rianfowler) [inline platform/engineering/design-docs/flag-for-manual-review.md]:
```suggestion
```

I don't think this strikes the right tone. We want to automate the easy stuff so VFS teams can focus their reviews on the harder problems. For the things the platform must enforce but can't automate, we're using this trigger.

File: platform/engineering/design-docs/flag-for-manual-review.md
Comment (rianfowler) [inline platform/engineering/design-docs/flag-for-manual-review.md]:
```suggestion
- Whether or not the branch it is running on has an associated PR
```

File: platform/engineering/manual-review-triggers.md
Comment (rianfowler) [inline platform/engineering/manual-review-triggers.md]: This is cool

**Review by rianfowler — APPROVED** [review]
cc @gunsch @johnpaulashenfelter

File: platform/engineering/code_review_guidelines.md
Comment (bkjohnson) [reply platform/engineering/code_review_guidelines.md]: Unfortunately [the API doesn't let us do that](https://developer.github.com/v3/pulls/review_requests/#create-a-review-request).  All we can do is say who we want to request a review from.

However, I included a section talking about other ways that we could enforce that they wait for a manual review.

File: platform/engineering/design-docs/flag-for-manual-review.md
Comment (bkjohnson) [reply platform/engineering/design-docs/flag-for-manual-review.md]: That's fair - I'll remove it.

File: platform/engineering/manual-review-triggers.md
Comment (gunsch) [inline platform/engineering/manual-review-triggers.md]:
```suggestion
We review logs to Sentry to ensure that any new logs will be useful and will not contain PII.

Examples:
- Don't log an entire request response if all that's important is an error code
- Don't log user-identifying information such as names, emails, or user IDs
```

File: platform/engineering/manual-review-triggers.md
Comment (gunsch) [inline platform/engineering/manual-review-triggers.md]:
```suggestion
Disabling eslint rules will be evaluated on a case-by-case basis.
```

File: platform/engineering/manual-review-triggers.md
Comment (gunsch) [inline platform/engineering/manual-review-triggers.md]:
```suggestion
We use Font Awesome as a dependency, which uses the `<i>` tag for adding icons. Sometimes an icon is used purely as decoration, but other times it is used to convey meaning to the user.

We review to ensure that whenever an icon is being used semantically, those semantics are also conveyed to a screen reader.
```

File: platform/engineering/design-docs/flag-for-manual-review.md
Comment (bkjohnson) [reply platform/engineering/design-docs/flag-for-manual-review.md]: I put the environment variables in the same doc with the list of triggers and linked to it.  I'm not convinced that that is the best place,  so that may change as I clean things up.

File: platform/engineering/design-docs/flag-for-manual-review.md
Comment (bkjohnson) [reply platform/engineering/design-docs/flag-for-manual-review.md]: I avoided that because that felt like part of the solution, and the template said to [keep that out of the background](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/engineering/design-docs/design-doc-template.md#background).  I'll try to phrase it in a way that keeps the emphasis on the problem and not the solution.

File: platform/engineering/flag-for-manual-review.md
Comment (gunsch) [inline platform/engineering/flag-for-manual-review.md]:
```suggestion
We will use a Node.js script that can be run from a job in CircleCI.
```

File: platform/engineering/flag-for-manual-review.md
Comment (gunsch) [inline platform/engineering/flag-for-manual-review.md]: From the [CircleCI docs](https://circleci.com/docs/2.0/env-vars/):

> Note: secret masking will only prevent the value of the environment variable from appearing in your build output. The value of the environment variable is still accessible to users debugging builds with SSH.

who has SSH access to these builds?

File: platform/engineering/manual-review-triggers.md
Comment (gunsch) [inline platform/engineering/manual-review-triggers.md]:
```suggestion
Disabling ESLint rules will be evaluated on a case-by-case basis.
```

File: platform/engineering/flag-for-manual-review.md
Comment (gunsch) [inline platform/engineering/flag-for-manual-review.md]: add @drorva to this list, he's asked to be on all design docs for now

```suggestion
**Approvers:** Rian Fowler \[ \], Andrew Gunsch \[X\], Dror Matalon \[ \]
```

**Review by gunsch — APPROVED** [review]
Design doc location: maybe back in the `design-docs` folder, and prepended with a date? we don't really have a convention yet for where these live, but the `engineering` directory is pretty convoluted already with a bunch of different files.

Couple small comments here, but otherwise this looks pretty good! adding my approval (but make sure to get others' also before submitting).

File: platform/engineering/flag-for-manual-review.md
Comment (bkjohnson) [reply platform/engineering/flag-for-manual-review.md]: My understanding is that anyone who has an SSH public key on Github will be able to connect. I re-ran a job with SSH and this is what I see:

image

So anytime we deliberately kick off a job with SSH access, we should be able to see whoever connects to it.

Comment (gunsch) [reply platform/engineering/flag-for-manual-review.md]: _anyone_ with an SSH public key? So, that sounds to me like the GitHub bot auth token can be accessed by anyone?

Comment (bkjohnson) [reply platform/engineering/flag-for-manual-review.md]: I found a more specific answer in [the docs](https://circleci.com/docs/2.0/ssh-access-jobs/):

> In order to SSH into a CircleCI build, the username must be one which has access to the project being built!

When I try to go to [the project website](https://app.circleci.com/pipelines/github/department-of-veterans-affairs/vets-website) in a private window it blocks me and forces me to authenticate with Github or Bitbucket.

I created a fake github account to test this and was able to login and view the project in Circle, but I can't even look at the project settings. I get errors when I try to follow the project, and I can't see the list of branches either.  The option to trigger an SSH build is disabled as well:

image

After seeing all of this, I think that project access in circle requires an account to be a part of the Github organization that the repo is associated with. I've looked around at various settings to try to get a more definitive answer, but I'm not an organization admin in Circle so there are certain things I can't see. I wonder if [this page](https://circleci.com/gh/organizations/department-of-veterans-affairs/settings#projects) would be helpful.

Comment (gunsch) [reply platform/engineering/flag-for-manual-review.md]: https://discuss.circleci.com/t/disable-ssh/31057 has a couple key things:

> Any user with write capabilities will also have SSH access. [...] any user with write permissions could theoretically also expose secrets via commands issues through the config.yml file.

Related question to investigate: how often do we hand out `write` permission on vets-website/vets-api, or do we generally keep people's access to `read`?

> A good first step is to ensure your deploy keys only have read access to your repos.

This made me wonder, what kinds of access does our bot token have? Can we keep that as minimal as possible? (read-only + posting comments?).

Other possible idea: it sounds like [contexts](https://circleci.com/docs/2.0/contexts/#restricting-a-context) might be useful here.

Comment (bkjohnson) [reply platform/engineering/flag-for-manual-review.md]: Do you know who gives `write` access to people?

Our bot token has the [`public_repo` scope](https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/#available-scopes).  I believe that this is the least permissions that we can give it in order for it to do its job, but it does include write access for our public repos.

> Limits access to public repositories. That includes read/write access to code, commit statuses, repository projects, collaborators, and deployment statuses for public repositories and organizations.

Here is a [relevant Slack thread](https://dsva.slack.com/archives/CJYRZK2HH/p1581623319018100?thread_ts=1581620875.013600&cid=CJYRZK2HH).

Contexts are something that I haven't explored deeply.  I'm interested to see how we could use them, but I don't have permissions to manage security groups so I'm a bit limited.

Comment (gunsch) [reply platform/engineering/flag-for-manual-review.md]: > Do you know who gives `write` access to people?

VA has an entire GitHub admin group. you can also talk to @ricetj about longer-term strategy for ownership on these repos, or (more immediately helpful) see if you can track down an admin for `vets-website` and `vets-api` to see how we've handed out write access so far. it's possible that most people just have `read` access and that this isn't a concern, but `write` access is typically needed for managing issues, which might mean we've handed it out more broadly.

> Our bot token has the [`public_repo` scope](https://developer.github.com/apps/building-oauth-apps/understanding-scopes-for-oauth-apps/#available-scopes). I believe that this is the least permissions that we can give it in order for it to do its job, but it does include write access for our public repos.

hmm, this set of information is making me more and more concerned 😬 

what does the bot need from `public_repo` that it can't get from `(no scope)`?

Comment (rianfowler) [reply platform/engineering/flag-for-manual-review.md]: FWIW- Write access is needed by a GitHub team for codeowners to trigger so everyone has write access to the repo. 

All of the FE group on VSP has admin access. 

Generally everyone has `write` access.

**Pull request merged**
